1756 dont display output from previous runs

mnowster opened the pull request:
This approach fixes to prevent `up` from displaying output of previous runs.

This approach is minimal viable changes to get the desired fix in with minimum disruption to rest of the structure of the code and ensures our users get a fix for 1.5 rather than having to wait until 1.6 when we'll tackle more of the logging features.

Fixes https://github.com/docker/compose/issues/1756

aanand commented:
Nice.

dnephin commented on compose/service.py:
```diff
@@ -448,11 +479,26 @@ def recreate_container(self,
             container.id,
-            '%s_%s' % (container.short_id, container.name))
+            '%s_%s' % (container.short_id, container.name)
+        )
```
I think the `Container` class would be a better place for this new function. It only uses `self.client` and `container` (which are both available in `Container`).

Maybe something like `def rename_to_tmp_name(self):` ?

dnephin commented on compose/service.py:
```diff
@@ -448,11 +479,26 @@ def recreate_container(self,
             else:
                 raise
```
I'm pretty sure we don't need the try/except anymore (it can be just a plain `container.stop(timeout=timeout)`).

The api returns a 304 for already stopped containers which doesn't raise an error. Looks like it was added over a year ago b573b87a92d12bad82a41377e8cbaa891ac0dfb9. The API has been fixed since then.

If you don't feel comfortable removing the try/except in this PR, maybe just add a TODO for now and we can remove it later?

dnephin commented on compose/cli/main.py:
```diff
@@ -565,17 +565,25 @@ def up(self, project, options):
+                list(containers),
+                monochrome=monochrome,
+                log_streams=list(log_streams)
+            )
```
It looks like you removed `build_log_printer` but then reproduced most of the logic inline and didn't remove the function?

Could this be moved back into `build_log_printer` ?

dnephin commented on compose/cli/log_printer.py:
```diff
@@ -36,10 +37,18 @@ def no_color(text):
+            for color_func, container in zip(color_funcs, self.containers):
+                generator_func = get_log_generator(container)
+                prefix = color_func(build_log_prefix(container, prefix_width))
+                yield generator_func(container, prefix, color_func)
```
I think a lot more of this code can be shared between the branches by making a small change to `get_log_generator()`. I was hoping to have a chance today to prototype that, but I've run out of time.

I think we can get it down to a single line in each if/else branch inside the for loop

dnephin commented on compose/project.py:
```diff
@@ -302,15 +303,17 @@ def up(self,
-        ]
+            containers += converged_results
+
+        return containers
```
This doesn't need to change, right? It's just an added parameter.  I think the compression version is more clear.

dnephin commented on compose/service.py:
```diff
@@ -395,49 +395,80 @@ def _containers_have_diverged(self, containers):
             self.start_container(container)
 
-            return [container]
+            return zip([container], [log_stream])
```
It would be really awesome if we could keep the same return value. I think a lot of the complexity in this change is having to accommodate this difference.

Since our log_stream comes from a container, I think we could make it a method on Container.

```python

class Container(object):
    def __init__(self, ...):
       ...
       self._log_stream = None


    def init_log_stream(self):
        if not container.has_api_logs:
            return
        self._log_stream = self.attach(stdout=True, stderr=True, stream=True)

    def get_log_stream(self):
        return self._log_stream
```

It's unfortunate that it wouldn't be immutable , but Container isn't immutable as it is, so I guess that's not so bad.

In this function I think we'll be able to restore the returning of just a list of containers, and it can do:

````python
    if not detached:
        container.init_log_stream()
````

dnephin commented on compose/service.py:
```diff
@@ -448,11 +479,26 @@ def recreate_container(self,
+
+    def recreate_container(self,
+                           container,
+                           timeout=DEFAULT_TIMEOUT):
```
Instead of duplicating this, with my proposed change to add init_log_stream/get_log_stream to Container, I think we can just add an `init_log_stream=False` kwarg here, and set it to `True` from `execute_convergence_plan()` to remove the need for duplication.

dnephin commented on compose/service.py:
```diff
@@ -395,27 +395,57 @@ def _containers_have_diverged(self, containers):
             container = self.create_container(do_build=do_build)
+
+            if container.has_api_logs and not detached:
+                container.attach_log_stream()
```
`attach_log_stream()` already checks for `has_api_logs()` (which I think is a good place to do it), so I think we can remove it from here and do just `if not detached:`

dnephin commented on compose/service.py:
```diff
@@ -395,27 +395,57 @@ def _containers_have_diverged(self, containers):
+                        container.attach_log_stream()
+
+                    log.info("Starting %s" % container.name)
+                    self.start_container(container)
```
If we add a flag `attach_logs` to `start_container_if_stopped()` which calls this. I think this would be:

```python

for container in containers:
    self.start_container_if_stopped(container, attach_logs=not detached)
```

which removes the need to duplicate anything

dnephin commented on compose/service.py:
```diff
@@ -395,27 +395,57 @@ def _containers_have_diverged(self, containers):
+                self.start_container(new_container)
+                container.remove()
+
+            return recreated_containers
```
I think we can do the same thing here.

If we add a flag `attach_logs` to `recreate_container()` which calls `attach_log_stream()`. I think this would be:

```python

return [
    self.recreate_container(container, timeout=timeout, attach_logs=not detached)
    for container in containers
]
```

which removes the need to duplicate anything

dnephin commented:
This is looking great, thanks for making the changes

Just a couple smaller things to remove some of the duplicate and I think this will be ready

mnowster commented:
@dnephin ok, I think I've addressed all feedback. I'm just running all tests locally now.

Yeh by introducing the `attach_logs` flag enabled me to remove that duplication so good call thank you :+1:

dnephin commented on compose/service.py:
```diff
@@ -395,27 +395,35 @@ def _containers_have_diverged(self, containers):
             return [container]
 
         elif action == 'recreate':
+            should_attach_logs = not detached
```
This could be moved out of the if/elif/elif block since it's used in all three branches, but super minor, not a blocker

mnowster replied:
~Good point, done :)

dnephin commented:
LGTM!

Thanks for getting this fixed, there will be many happy users!

mnowster commented:
@dnephin can confirm all tests pass on my local machine btw :sparkles:

dnephin commented:
tests pass locally

thaJeztah commented:
\o/ thanks @mnowster !

bfirsh commented:
woot!

szuliq commented:
Not sure if it was intended but attaching to a detached container still duplicates logs. 

Using `docker-compose==1.5.2` the following

    docker-compose up -d xxx
    docker-compose up xxx

prints all the logs from the backlog of `xxx` whereas

    docker-compose up xxx

when `xxx` is not running in the background does not.

When I say backlog, I mean since the container was created, not since it was last run "interactively".

I think I know why it's that way. Attaching probably is equivalent of `docker-compose logs xxx` with `Ctrl+C` binding.

Nevertheless my guts tell me the calls to `docker-compose up xxx` should behave in the same way regardless of the `xxx` running in the background or not.

Pomax commented:
Seeing a complete log-since-creation scrolling by again, was there a regression that nullified this change?
